gnocchi/chef.py

```diff
@@ -51,6 +53,80 @@ def __init__(self, coord, incoming, index, storage):
         self.index = index
         self.storage = storage
 
+    def resource_ended_at_normalization(self, metric_inactive_after):
+        """Marks resources as ended at if needed.
+
+        This method will check all metrics that have not received new
+        datapoints after a given period. The period is defined by
+        'metric_inactive_after' parameter. If all metrics of resource are in
+        inactive state, we mark the ended_at field with a timestmap. Therefore,
+        we consider that the resource has ceased existing.
+
+        In this process we will handle only metrics that are considered as
+        inactive, according to `metric_inactive_after` parameter. Therefore,
+        we do not need to lock these metrics while processing, as they are
+        inactive, and chances are that they will not receive measures anymore.
+        Moreover, we are only touching metadata, and not the actual data.
+        """
+
+        momment_now = utils.utcnow()
+        momment = momment_now - datetime.timedelta(
+            seconds=metric_inactive_after)
+
+        inactive_metrics = self.index.list_metrics(
+            attribute_filter={"<": {
+                "last_measure_timestamp": momment}},
+            resource_policy_filter={"==": {"ended_at": None}}
+        )
+
+        LOG.debug("Inactive metrics found for processing: [%s].",
+                  inactive_metrics)
+
+        metrics_by_resource_id = {}
+        for metric in inactive_metrics:
+            resource_id = metric.resource_id
+            if metrics_by_resource_id.get(resource_id) is None:
+                metrics_by_resource_id[resource_id] = []
+
+            metrics_by_resource_id[resource_id].append(metric)
+
+        for resource_id in metrics_by_resource_id.keys():
+            if resource_id is None:
+                LOG.debug("We do not need to process inactive metrics that do "
+                          "not have resource. Therefore, these metrics [%s] "
+                          "will be considered inactive, but there is nothing "
+                          "else we can do in this process.",
+                          metrics_by_resource_id[resource_id])
+                continue
+            resource = self.index.get_resource(
+                "generic", resource_id, with_metrics=True)
+            resource_metrics = resource.metrics
+            resource_inactive_metrics = metrics_by_resource_id.get(resource_id)
+
+            all_metrics_are_inactive = True
+            for m in resource_metrics:
```

Comment:
it wouldnt be simple if we just compare the size of resource_metrics and resource_inactive metrics? if they are the same, then we can clean, if not, then we ignore it.

Reply:
It feels risky. I find it more secure to actually check if the metrics are indeed the same.

Reply:
I can understand your point about safety, but if you are collecting both `resource_metrics` and `inactive_metrics` from the same source (same DB table) I cant see a scenario where matching both sizes wouldn't fit this goal. Maybe I am not seeing something that you saw. 

It was just a suggestion to reduce code complexity and improve a little the runtime, nothing critical.

Reply:
Yes, I load them from the DB. However, they are loaded in different moments in the code. Therefore, there is no 100% that there is no other manipulation happening in between the different loads/processing. I mean, it is very unlikely, but possible.

Reply:
I see, but both solutions would fail in some edge cases (like new measures for inactive_metrics while this process is running). 

The size comparing solution would fail also if someone deletes a metric and creates a new metric while this process is running (which is less likely than new measures coming), but in the end, it would only mark this resource as ended, not delete it indeed, and as it receives any new measurements, it would get back to active, right? I guess you did it to deal with such edge cases.

I see no problem with your solution, it is good to me as well, I think both solutions would work fine, I just thought that reducing complexity would be something nice. :slightly_smiling_face:

Reply:
exactly that. And even if the resource is marked as "deleted", as soon as a metric receives new datapoints, it is "undeleted".